File: js/day20.js
window.AdventOfCode.Day20 = function( input )
{
	input = +input;
	
	var i, j;
	var n = input / ( input / 1E6 );
	var houses = {};
	var housesPartTwo = {};
	
	for( i = 1; i < n; i++ )
	{
		for( j = i; j < n; j += i )
		{
			houses[ j ] = ( houses[ j ] || 0 ) + i * 10;
			
			if( houses[ j ] >= input )
			{
				break;
			}
		}
		
		var lazyElfs = 0;
		for( j = i; lazyElfs < 50; j += i )
		{
			housesPartTwo[ j ] = ( housesPartTwo[ j ] || 0 ) + i * 11;
			
			if( housesPartTwo[ j ] >= input )
			{
				break;
			}
			
			lazyElfs++;
		}
	}
	
	for( i = 1; i < input; i++ )
	{
		if( houses[ i ] >= input )
		{
			break;
		}
	}
	
	for( j = 1; j < input; j++ )
	{
		if( housesPartTwo[ j ] >= input )
		{
			break;
		}
	}
	
	return [ i, j ];
};

File: 2024/Answers/Solutions/Day20.cs
using System;
using System.Collections.Generic;

namespace AdventOfCode;

[Answer(20)]
public class Day20 : IAnswer
{
	record struct Step(int X, int Y, int Cost);

	public Solution Solve(string input)
	{
		var part1 = 0;
		var part2 = 0;
		var grid = new ReadOnlyGrid(input, '#');
		var finish = grid.IndexOf('E');
		var position = grid.IndexOf('S');
		var last = position;
		var cost = 1;
		var path = new List<Step>(1024 * 10)
		{
			new (position.X, position.Y, 0)
		};

		while (position != finish)
		{
			foreach (var delta in Coord.Directions)
			{
				var pos = position + delta;

				if (pos != last && grid[pos] != '#')
				{
					last = position;
					position = pos;
					path.Add(new Step(pos.X, pos.Y, cost++));
					break;
				}
			}
		}

		for (var i = 0; i < path.Count; i++)
		{
			var step = path[i];

			for (var j = i + 1; j < path.Count; j++)
			{
				var stepEnd = path[j];
				var distance = Math.Abs(stepEnd.Y - step.Y) + Math.Abs(stepEnd.X - step.X);

				if (distance <= 20 && stepEnd.Cost - step.Cost - distance >= 100)
				{
					part2++;

					if (distance == 2)
					{
						part1++;
					}
				}
			}
		}

		return new(part1.ToString(), part2.ToString());
	}
}
